File: src/main/java/david/java/flink/watermark/BoundedOutOfOrdernessGenerator.java
package david.java.flink.watermark;

import david.java.flink.pojo.MyEvent;
import org.apache.flink.api.common.eventtime.Watermark;
import org.apache.flink.api.common.eventtime.WatermarkGenerator;
import org.apache.flink.api.common.eventtime.WatermarkOutput;

/**
 * @Description: a periodic watermark work on event time
 * ExecutionConfig.setAutoWatermarkInterval(...)   define watermark interval
 * @Author: Zhaodawei
 * @Date: Create in 下午5:43 2020/12/18
 */
public class BoundedOutOfOrdernessGenerator implements WatermarkGenerator<MyEvent> {

    private final long maxOutOfOrderness = 3500;

    private long currentMaxTimestamp;

    @Override
    public void onEvent(MyEvent myEvent, long eventTimestamp, WatermarkOutput watermarkOutput) {
        currentMaxTimestamp = Math.max(currentMaxTimestamp, eventTimestamp);
    }

    @Override
    public void onPeriodicEmit(WatermarkOutput output) {
        output.emitWatermark(new Watermark(currentMaxTimestamp - maxOutOfOrderness -1));
    }
}

File: src/main/scala/david/scala/flink/watermark/BoundedOutOfOrdernessGenerator.scala
package david.scala.flink.watermark

import david.java.flink.pojo.MyEvent
import org.apache.flink.api.common.eventtime.{Watermark, WatermarkGenerator, WatermarkOutput}


/**
 * @Description: there is a mistake about BoundedOutOfOrdernessGenerator on offical  webset
 * @Author: Zhaodawei
 * @Date: Create in 下午12:43 2020/12/21 
 */
class BoundedOutOfOrdernessGenerator extends WatermarkGenerator[MyEvent] {
//class BoundedOutOfOrdernessGenerator extends AssignerWithPeriodicWatermarks[MyEvent] {
    val maxOutOfOrderness = 3500L
    var currentMaxTimestamp: Long = _

    override def onEvent(t: MyEvent, l: Long, watermarkOutput: WatermarkOutput): Unit = {
        currentMaxTimestamp = Math.max(l, currentMaxTimestamp)
    }

    override def onPeriodicEmit(watermarkOutput: WatermarkOutput): Unit = {
        watermarkOutput.emitWatermark(new Watermark(currentMaxTimestamp - maxOutOfOrderness - 1))
    }
}
